llvm/tools/sycl-post-link/SpecConstants.h

```diff
@@ -11,20 +11,21 @@
 // constant operations. The spec constant IDs are symbolic before linkage to
 // make separate compilation possible. After linkage all spec constants are
 // available to the pass, and it can assign consistent integer IDs.
-//
-// The pass is used w/o a pass manager currently, but the interface is based on
-// the standard Module pass interface to move it around easier in future.
 //===----------------------------------------------------------------------===//
 
 #pragma once
 
-#include "llvm/ADT/StringMap.h"
 #include "llvm/IR/Module.h"
 #include "llvm/IR/PassManager.h"
 
-using namespace llvm;
+#include <map>
+#include <vector>
+
+namespace llvm {
+class StringRef;
```

Comment:
The idea is to minimize amount of includes and therefore speedup compilation.

I agree that in this particular case it probably doesn't make any difference, as the header is only used twice (by `SpecConstants.cpp` and `sycl-post-link.cpp`), but nevertheless, I just applied this approach here.

Reply:
I see. llvm::Module and llvm::PreservedAnalyses would have to also be handled in similar way, and this approach does not scale. But consider this a nit, I don't see any harm in this case.

Reply:
> and this approach does not scale.

Yes, it indeed requires some effort

> I see. llvm::Module and llvm::PreservedAnalyses would have to also be handled in similar way, 

We can't do so, because they are passed by reference and therefore required to be fully defined.

Note, this is a common approach in LLVM project: llvm/llvm-project@43d6f9a87678bde0ba662c6486f2e1a67c28681a, llvm/llvm-project@3caa03ec510d77cfc1bbb00b22d72d499183f6c6

Reply:
Please, read LLVM's documentation - coding standards in particular.
https://llvm.org/docs/CodingStandards.html#minimal-list-of-includes

> #include hurts compile time performance. Don’t do it unless you have to, especially in header files.
> 
> But wait! Sometimes you need to have the definition of a class to use it, or to inherit from it. In these cases go ahead and #include that header file. Be aware however that there are many cases where you don’t need to have the full definition of a class. If you are using a pointer or reference to a class, you don’t need the header file. If you are simply returning a class instance from a prototyped function or method, you don’t need it. In fact, for most cases, you simply don’t need the definition of a class. And not #includeing speeds up compilation.

Reply:
> Please, read LLVM's documentation - coding standards in particular.
https://llvm.org/docs/CodingStandards.html#minimal-list-of-includes

Thanks! I was looking for it as well. It seems like my understanding about references is wrong, I will probably send some follow-up refactoring patches

Reply:
Right, then this still holds:

> llvm::Module and llvm::PreservedAnalyses would have to also be handled in similar way

This, for example, compiles just fine, as there is no real use of the object:
```cpp
class Foo;
void foo(Foo&);
```